File: trunk/lucette2/src/main/java/battle/csn/lucette/game/structure/MultiTab.java
package battle.csn.lucette.game.structure;

import java.util.ArrayList;
import java.util.List;

import org.apache.commons.lang3.ArrayUtils;


/**
 * Structure de donnée multidimensionelle pour enregistrer la valeurs des cases
 * d'un jeu à n dimension
 * @author Jeremy
 * 
 * @param <T> type d'éléments enregistrés
 */
public class MultiTab<T> implements ICases<T>
{
    private int _deep;
    private T _value;
    private List<MultiTab<T>> _subTab;
    private int[] _dimensions;

    /**
     * Get the sub tab
     * @param index the index of the sub tab
     * @return the tab
     */
    private MultiTab<T> getSubTab( int index )
    {
        return _subTab.get( index );
    }

    public MultiTab( )
    {
        this._deep = 1;
    }

    @Override
    public int[] getSizes( )
    {
        return _dimensions;
    }

    @Override
    public T get( int... dimensions )
    {
        if ( dimensions.length == 0 )
        {
            return _value;
        }
        else
        {
            return this.getSubTab( dimensions[0] ).get( ArrayUtils.remove( dimensions, 0 ) );
        }
    }

    @Override
    public void set( T value, int... dimensions )
    {
        if ( dimensions.length == 0 )
        {
            _value = value;
        }
        else
        {
            this.getSubTab( dimensions[0] ).set( value, ArrayUtils.remove( dimensions, 0 ) );
        }
    }

    @Override
    public void updateSizes( int... dimensions )
    {
        _dimensions = dimensions;
        _deep = dimensions.length;
        if ( dimensions.length > 0 )
        {
            int[] dimensionsUpdate = ArrayUtils.remove( dimensions, 0 );
            _subTab = new ArrayList<MultiTab<T>>( );
            for ( int i = 0; i < dimensions[0]; i++ )
            {
                _subTab.add( new MultiTab<T>( ) );
                _subTab.get( i ).updateSizes( dimensionsUpdate );
            }
        }
        else
        {
            this._value = null;
        }
    }

    /**
     * @return the _deep
     */
    public int getDeep( )
    {
        return _deep;
    }

    @Override
    public String toString( )
    {
        String result = "";
        if ( _deep == 1 )
        {
            for ( int i = 0; i < this._dimensions[0]; i++ )
            {
                result += this.getSubTab( i ).get( );
            }
        }
        return result;
    }
}

File: lucette2/src/main/scala/battle/csn/lucette2/game/structure/MultiTab.scala
package battle.csn.lucette2.game.structure

import org.apache.commons.lang3.ArrayUtils
import scala.collection.mutable.MutableList

class  MultiTab[T] extends ICases[T] {
  var deep: Int = _
  var value: Option[T] = _
  var subTab= new MutableList[MultiTab[T]]()
  var dimensions: Seq[Int] = _
  var move:Move=_

  def MultiTab() = this.deep = 1

  def getSizes() = dimensions

  def get(dimensions: Seq[Int]) = {
    if (dimensions.length == 0) {
      value 
    } else {
      subTab(dimensions(0)).get(dimensions.drop(1))
    }
  }
  
  def set(value: T, dimensions: Seq[Int]) = {
    if (dimensions.length == 0) {
      this.value = Some(value) 
     }
    else {
      subTab(dimensions(0)).set(value, dimensions.drop(1))
    }
  }

  @Override
  def updateSizes(dimensions: Seq[Int]) =
    {
      this.dimensions = dimensions;
      this.deep = dimensions.length;
      if (dimensions.length > 0) {
        var dimensionsUpdate = dimensions.drop(1);
        var i = 0
        for (i <- 0 to dimensions(0)-1) {
          var t = new MultiTab[T]()
          t.updateSizes(dimensionsUpdate);
          this.subTab += t
        }
      } else {
        this.value = None
      }
    }
  
  def size()=subTab.size

  /**
   * @return the _deep
   */
  def getDeep = deep

  override def toString() =
    {
      var result = "";
      if (deep == 1) {
        var i = 0
        for (i <- 0 to dimensions(0)) {
          result += subTab(i).get(Seq());
        }
      }
      result;
    }
}